### PR TITLE
Config validation

### LINKED ISSUE #2088: Automatically validate blockchain config
## Problem

With the PR #2053 we have a simpler config which allows a user to specify the network (or chain_id) and url. 
The combination is not verified which might lead to problems during runtime, e.g.  one could specify `mainnet` as the network but provide a url aiming at `testnet`. 

## Goal

Cnd should verify that the provided config is valid and validate that cnd is connected to the desired blockchain nodes

### PR DESCRIPTION
**Pull request by rishflab**
This PR checks to see if the bitcoin node network id and ethereum node chain id matches the id's specified in the config file. The check is performed at at startup. A warning is logged if the request to retrieve the id fails. Cnd will abort if the check finds that the actual id does not match the specified id.

Closes #2088.

### CODE REVIEW

File: cnd/src/config/validation.rs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+#[derive(Debug)]
+pub enum NetworkValidationResult<T> {
+    Valid,
+    Invalid {
+        connected_network: T,
+        specified_network: T,
+    }
+}
+
+pub async fn validate_ethereum_chain_id(connection: &Web3Connector, specified: ChainId) -> Result<NetworkValidationResult<ChainId>, anyhow::Error> {
```
Comment (rishflab) [inline cnd/src/config/validation.rs]: I can replace this and the function below with a generic version which takes the function to fetch the network/chain id as a parameter

**Comment by thomaseizinger** [comment]
You didn't use a closing keyword in your PR description, so the issue is not going to be closed automatically. You might want to fix that :)

File: cnd/src/main.rs [hunk 2]
```diff
@@ -162,3 +176,13 @@ fn dump_config(settings: Settings) -> anyhow::Result<()> {
...
+
+/// Throws an error if the connected connected blockchain network is invalid
+fn handle_invalid_network_validation<T: std::fmt::Debug>(validation: NetworkValidationResult<T>) -> anyhow::Result<(), ConfigValidationError<T>> {
+    match validation {
+        NetworkValidationResult::Invalid { connected_network: connected, specified_network: specified } => {
+            Err(ConfigValidationError::ConnectedNetworkDoesNotMatchSpecified { connected_network: connected, specified_network: specified })
+        },
+        NetworkValidationResult::Valid => Ok(())
+    }
+}
```
Comment (bonomat) [inline cnd/src/main.rs]: Missing new line :) 

```suggestion
}

```

File: cnd/src/btsieve/bitcoin/mod.rs [hunk 3]
```diff
@@ -1,9 +1,22 @@
 mod bitcoind_connector;
 mod cache;
 
+<<<<<<< HEAD
```
Comment (tcharding) [inline cnd/src/btsieve/bitcoin/mod.rs]: Please remove these merge conflict lines.

File: cnd/src/btsieve/ethereum/web3_connector.rs [hunk 4]
```diff
@@ -61,6 +61,41 @@ impl BlockByHash for Web3Connector {
...
+
+    let chain_id = match response {
+        JsonRpcResponse::Success { result } => result,
+        JsonRpcResponse::Error { code, message } => {
+            tracing::warn!("net_version request failed with {}: {}", code, message);
+            return Err(anyhow::anyhow!(
+                "net_version request failed with {}: {}",
+                code,
+                message
+            ));
```
Comment (tcharding) [inline cnd/src/btsieve/ethereum/web3_connector.rs]: No need to log a warning and return an error with the same information, IMO.

File: cnd/src/config/validation.rs [hunk 5]
```diff
@@ -9,17 +9,14 @@ use crate::btsieve::{
...
-        specified_network: T,
-    },
     #[error("Connected network does not match network specified in settings (expected {connected_network:?}, got {specified_network:?})")]
     ConnectedNetworkDoesNotMatchSpecified {
         connected_network: T,
         specified_network: T,
     }
 }
 
+#[derive(Debug)]
```
Comment (tcharding) [inline cnd/src/config/validation.rs]: This should be in the commit that introduces this struct.

File: cnd/src/main.rs [hunk 6]
```diff
@@ -72,6 +72,7 @@ fn main() -> anyhow::Result<()> {
             BITCOIN_BLOCK_CACHE_CAPACITY,
         )
     };
+<<<<<<< HEAD
```
Comment (tcharding) [inline cnd/src/main.rs]: Please remove this.

File: cnd/src/config/validation.rs [hunk 7]
```diff
@@ -25,6 +26,19 @@ pub enum NetworkValidationResult<T> {
     }
 }
 
+pub async fn validate_connected_network<T: std::cmp::PartialEq, U, F: Fn(U) -> Box<dyn Future<Item=NetworkValidationResult<T>, Error=anyhow::Error>>> (
```
Comment (tcharding) [inline cnd/src/config/validation.rs]: Are you running `rustfmt` in your editor?  It seems strange to see this `=` with no spaces around it?

File: cnd/src/btsieve/ethereum/web3_connector.rs [hunk 8]
```diff
@@ -61,35 +62,27 @@ impl BlockByHash for Web3Connector {
...
+    //     jsonrpc::Response::Success { result } => result,
+    //     jsonrpc::Response::Error { code, message } => {
+    //         tracing::warn!("net_version request failed with {}: {}", code, message);
+    //         return Err(anyhow::anyhow!(
+    //             "net_version request failed with {}: {}",
+    //             code,
+    //             message
+    //         ));
+    //     }
+    // };
```
Comment (tcharding) [inline cnd/src/btsieve/ethereum/web3_connector.rs]: Please remove this.

**Comment by tcharding** [comment]
Added no-mergify because to give you a chance to remove all the merge conflict indicators (or squash the PR into a single commit).  Please do not merge with `<<<<<<<` lines in there.

**Comment by rishflab** [comment]
> Nicely done.
> 
> * What were your thoughts on testing your implementations?
> * Imho we could add a note to `changelog.md` pointing out that cnd will report an error if the settings and nodes don't match

Had a discussion with Thomas regarding testing this. The happy path is already tested by our current e2e tests. To test the sad path I would have to create a new e2e test that starts a full bitcoin or ethereum node with the wrong network or chain id. We decided it wasn't worth it.  Did not think it was worth having unit tests given the simplicity of the logic being tested. 

I will update the changelog before merging.

**Pull request merged**